### ISSUE TITLE
Implications of `Arc<K>: Borrow<Q>`

### ISSUE THREAD

**Issue by tatsuya6502** [moka-rs/moka#166]
### Discussed in https://github.com/moka-rs/moka/discussions/163

<div type='discussions-op-text'>

<sup>Originally posted by **JayKickliter** July  9, 2022</sup>
For my particular use case I'm finding `moka::sync::Cache` requirement of `Arc<K>: Borrow<Q>` very restrictive, as my keytype is somewhat expensive to create. Is this an absolute requirement for this kind of cache or an implementation detail?

## Expected behavior

Using `HashMap` as an example, I'd expect to be able to querey using a `&[u8]` when the key type is `Vec<u8>`.

```rust
    use std::collections::HashMap;
    let mut map: HashMap<Vec<u8>, ()> = HashMap::new();
    map.insert(vec![1_u8], ());
    map.contains_key([1_u8].as_slice());
```

## Failing example, simplified

```rust
    use moka::sync::Cache;
    let cache: Cache<Vec<u8>, ()> = Cache::new(1);
    cache.insert(vec![1_u8], ());
    cache.contains_key([1_u8].as_slice());
```

```
error[E0277]: the trait bound `Arc<Vec<u8>>: Borrow<[u8]>` is not satisfied
   --> src/main.rs:5:11
    |
5   |     cache.contains_key([1_u8].as_slice());
    |           ^^^^^^^^^^^^ the trait `Borrow<[u8]>` is not implemented for `Arc<Vec<u8>>`
    |
    = help: the trait `Borrow<T>` is implemented for `Arc<T>`
note: required by a bound in `moka::sync::Cache::<K, V, S>::contains_key`
   --> /Users/jay/.cargo/registry/src/github.com-1ecc6299db9ec823/moka-0.9.0/src/sync/cache.rs:880:17
    |
880 |         Arc<K>: Borrow<Q>,
    |                 ^^^^^^^^^ required by this bound in `moka::sync::Cache::<K, V, S>::contains_key`
```

**EDIT**: I now wonder if this could be considered a bug after noticing the following.

This also seems to imply that the docs [`contains_key`]  are incorrect:

> The key may be any borrowed form of the cache’s key type,

[`contains_key`]: https://docs.rs/moka/latest/moka/sync/struct.Cache.html#method.contains_key</div>

**Comment by tatsuya6502** [comment]
I found a solution for `sync` and `future` caches. I am working on it here:

- #167
    - [Added unit test](https://github.com/moka-rs/moka/pull/167/files#diff-3eee177f538ae36c5dc747eb92f8b63b850012a55c7289c3f660d11971fb211eR2952)

If it goes well, I will publish Moka v0.9.1 with the solution.

**Comment by tatsuya6502** [comment]
I am closing this issue.

- Resolved for the `sync` and `future` caches via #167:
    - I published Moka [v0.9.1](https://crates.io/crates/moka/0.9.1) with the this fix.
- Won't fix for the `unsync` and experimental `dash` caches:
    - I could not find a way to fix it for them without modifying the source codes of `std::collections::HashMap` (for the `unsync` cache) and `dashmap::DashMap` (for the `dash` cache)
